Sources/SentryCrash/Recording/SentryCrashBinaryImageCache.c

```diff
@@ -0,0 +1,156 @@
+    = &_dyld_register_func_for_add_image;
+static SentryRegisterFunction _sentry_register_func_for_remove_image
+    = &_dyld_register_func_for_remove_image;
+
+void
+sentry_setRegisterFuncForAddImage(SentryRegisterFunction addFunction)
+{
+    _sentry_register_func_for_add_image = addFunction;
+}
+
+void
+sentry_setRegisterFuncForRemoveImage(SentryRegisterFunction removeFunction)
+{
+    _sentry_register_func_for_remove_image = removeFunction;
+}
+
+void
+sentry_resetFuncForAddRemoveImage(void)
+{
+    _sentry_register_func_for_add_image = &_dyld_register_func_for_add_image;
+    _sentry_register_func_for_remove_image = &_dyld_register_func_for_remove_image;
+}
+
+#    define sentry_dyld_register_func_for_add_image(CALLBACK)                                      \
+        _sentry_register_func_for_add_image(CALLBACK);
+#    define sentry_dyld_register_func_for_remove_image(CALLBACK)                                   \
+        _sentry_register_func_for_remove_image(CALLBACK);
+
+#else
+#    define sentry_dyld_register_func_for_add_image(CALLBACK)                                      \
+        _dyld_register_func_for_add_image(CALLBACK)
+#    define sentry_dyld_register_func_for_remove_image(CALLBACK)                                   \
+        _dyld_register_func_for_remove_image(CALLBACK)
+#endif
+
+typedef struct SentryCrashBinaryImageNode {
+    SentryCrashBinaryImage image;
+    bool available;
+    struct SentryCrashBinaryImageNode *next;
+} SentryCrashBinaryImageNode;
+
+static SentryCrashBinaryImageNode rootNode = { 0 };
+static SentryCrashBinaryImageNode *tailNode = NULL;
+static pthread_mutex_t binaryImagesMutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+binaryImageAdded(const struct mach_header *header, intptr_t slide)
+{
+    if (tailNode == NULL) {
+        return;
+    }
+
+    Dl_info info;
+    if (!dladdr(header, &info) || info.dli_fname == NULL) {
+        return;
+    }
+
+    SentryCrashBinaryImage binaryImage = { 0 };
+    if (!sentrycrashdl_getBinaryImageForHeader(
+            (const void *)header, info.dli_fname, &binaryImage)) {
+        return;
+    }
+
+    SentryCrashBinaryImageNode *newNode = malloc(sizeof(SentryCrashBinaryImageNode));
+    newNode->available = true;
+    newNode->image = binaryImage;
+    newNode->next = NULL;
+
+    pthread_mutex_lock(&binaryImagesMutex);
+    tailNode->next = newNode;
+    tailNode = tailNode->next;
+    pthread_mutex_unlock(&binaryImagesMutex);
+}
+
+static void
+binaryImageRemoved(const struct mach_header *header, intptr_t slide)
+{
+    SentryCrashBinaryImageNode *nextNode = &rootNode;
+
+    while (nextNode != NULL) {
+        if (nextNode->image.address == (uint64_t)header) {
+            nextNode->available = false;
+            break;
+        }
+        nextNode = nextNode->next;
+    }
+}
+
+void
+sentrycrashbic_iterateOverImages(sentrycrashbic_imageIteratorCallback callback, void *context)
+{
+    SentryCrashBinaryImageNode *nextNode = &rootNode;
```

Comment:
`m`: Please add a comment on why we don't need the mutex here and also for `binaryImageRemoved`.